File: ingestion-service/src/main/scala/datainsider/data_cook/domain/persist/EmailConfiguration.scala
package datainsider.data_cook.domain.persist

import com.twitter.finatra.validation.constraints.NotEmpty
import datainsider.client.exception.BadRequestError
import datainsider.data_cook.util.StringUtils

/**
  * @author tvc12 - Thien Vi
  * @created 03/15/2022 - 1:46 PM
  */

case class EmailConfiguration(
    @NotEmpty receivers: Seq[String],
    cc: Seq[String] = Seq.empty,
    bcc: Seq[String] = Seq.empty,
    subject: String,
    fileName: String,
    content: Option[String] = None,
    displayName: Option[String] = None
) extends ActionConfiguration {
  /**
   * Valid a action is correct
   */
  override def validate(): Unit = {
    val invalidReceivers: Seq[String] =  receivers.filterNot(email => StringUtils.isEmailFormat(email))
    val invalidCc: Seq[String] =  cc.filterNot(email => StringUtils.isEmailFormat(email))
    val invalidBcc: Seq[String] =  bcc.filterNot(email => StringUtils.isEmailFormat(email))
    if (invalidReceivers.nonEmpty || invalidCc.nonEmpty || invalidBcc.nonEmpty) {
      val invalidMails: Seq[String] = invalidReceivers ++ invalidCc ++ invalidBcc
      throw BadRequestError(s"invalid email address format: ${invalidMails}")
    }
  }
}

File: rocket-bi-web/di_core/DataCook/Domain/ETL/EmailConfiguration/EmailConfiguration.ts
export class EmailConfiguration {
  displayName: string;
  receivers: string[];
  cc: string[];
  bcc: string[];
  subject: string;
  content: string;
  fileName: string;

  constructor(displayName: string, receivers: string[], subject: string, content: string, fileName: string, cc: string[] = [], bcc: string[] = []) {
    this.displayName = displayName;
    this.receivers = receivers;
    this.subject = subject;
    this.content = content;
    this.fileName = fileName;
    this.bcc = bcc;
    this.cc = cc;
  }

  static fromObject(obj: EmailConfiguration): EmailConfiguration {
    return new EmailConfiguration(obj.displayName, obj.receivers, obj.subject, obj.content, obj.fileName, obj.bcc, obj.cc);
  }

  static default(fileName: string) {
    return new EmailConfiguration('', [], '', '', fileName, [], []);
  }
}
